File: scripts/generate_icon.py
#!/usr/bin/env python3
"""Draw the Qota Dock icon: three quota bars plus a check badge."""

from __future__ import annotations

from pathlib import Path

from PIL import Image, ImageDraw, ImageFilter

SIZE = 1024
BG = (26, 29, 36, 255)
TRACK = (54, 59, 70, 255)
TEAL = (45, 201, 176, 255)
AMBER = (240, 163, 58, 255)
CORAL = (232, 108, 116, 255)
BADGE_FILL = (20, 46, 44, 255)


def _circle(draw: ImageDraw.ImageDraw, cx: float, cy: float, r: float, fill) -> None:
    draw.ellipse((cx - r, cy - r, cx + r, cy + r), fill=fill)


def _capsule(draw: ImageDraw.ImageDraw, x: float, y: float, w: float, h: float, fill) -> None:
    draw.rounded_rectangle((x, y, x + w, y + h), radius=h / 2, fill=fill)


def _draw_mark(layer: Image.Image, alpha: int) -> None:
    draw = ImageDraw.Draw(layer, "RGBA")
    fill = (*TEAL[:3], alpha)
    track = (*TRACK[:3], alpha)
    badge = (*BADGE_FILL[:3], alpha)
    amber = (*AMBER[:3], alpha)
    coral = (*CORAL[:3], alpha)

    bar_h = 96
    gap = 42
    bar_w = 548
    left = 118
    top = (SIZE - (bar_h * 3 + gap * 2)) / 2
    fills = ((TEAL[:3] + (alpha,), 0.70), (amber, 0.46), (coral, 0.84))

    for i, (color, used) in enumerate(fills):
        y = top + i * (bar_h + gap)
        _capsule(draw, left, y, bar_w, bar_h, track)
        _capsule(draw, left, y, max(bar_h, bar_w * used), bar_h, color)

    cx, cy, r = 778, 548, 168
    _circle(draw, cx, cy, r, fill)
    _circle(draw, cx, cy, r - 26, badge)

    # Thick rounded check
    check = [
        (cx - 78, cy + 8),
        (cx - 18, cy + 68),
        (cx + 86, cy - 58),
    ]
    draw.line(check, fill=fill, width=46, joint="curve")
    for x, y in check:
        _circle(draw, x, y, 23, fill)


def render() -> Image.Image:
    canvas = Image.new("RGBA", (SIZE, SIZE), BG)

    silhouette = Image.new("RGBA", (SIZE, SIZE), (0, 0, 0, 0))
    _draw_mark(silhouette, 255)
    alpha = silhouette.split()[-1]
    dark = Image.new("RGBA", (SIZE, SIZE), (0, 0, 0, 0))
    dark.putalpha(alpha.point(lambda a: int(a * 0.45)))
    dark = dark.filter(ImageFilter.GaussianBlur(22))
    canvas.alpha_composite(dark, (0, 18))

    mark = Image.new("RGBA", (SIZE, SIZE), (0, 0, 0, 0))
    _draw_mark(mark, 255)
    canvas.alpha_composite(mark)
    return canvas


def main() -> None:
    root = Path(__file__).resolve().parent.parent
    dest = root / "Qota" / "Assets.xcassets" / "AppIcon.appiconset" / "AppIcon.png"
    dest.parent.mkdir(parents=True, exist_ok=True)
    image = render().convert("RGB")
    image.save(dest, "PNG", optimize=True)
    print(f"wrote {dest} ({dest.stat().st_size} bytes)")


if __name__ == "__main__":
    main()

File: scripts/generate-icon.swift
#!/usr/bin/env swift
import AppKit
import Foundation

let size = 1024.0
let dest = URL(fileURLWithPath: CommandLine.argc > 1
    ? CommandLine.arguments[1]
    : FileManager.default.currentDirectoryPath
        + "/Qota/Assets.xcassets/AppIcon.appiconset/AppIcon.png")

func color(_ r: CGFloat, _ g: CGFloat, _ b: CGFloat, _ a: CGFloat = 1) -> NSColor {
    NSColor(srgbRed: r, green: g, blue: b, alpha: a)
}

let bg = color(26 / 255, 29 / 255, 36 / 255)
let track = color(54 / 255, 59 / 255, 70 / 255)
let teal = color(45 / 255, 201 / 255, 176 / 255)
let amber = color(240 / 255, 163 / 255, 58 / 255)
let coral = color(232 / 255, 108 / 255, 116 / 255)
let badgeFill = color(20 / 255, 46 / 255, 44 / 255)

func capsule(_ rect: CGRect) -> NSBezierPath {
    NSBezierPath(roundedRect: rect, xRadius: rect.height / 2, yRadius: rect.height / 2)
}

let image = NSImage(size: NSSize(width: size, height: size))
image.lockFocus()
bg.setFill()
NSBezierPath(rect: NSRect(x: 0, y: 0, width: size, height: size)).fill()

let barH: CGFloat = 96
let gap: CGFloat = 42
let barW: CGFloat = 548
let left: CGFloat = 118
let groupH = barH * 3 + gap * 2
let top = (size - groupH) / 2
let fills: [(NSColor, CGFloat)] = [(teal, 0.70), (amber, 0.46), (coral, 0.84)]

for (i, item) in fills.enumerated() {
    let y = size - top - CGFloat(i + 1) * barH - CGFloat(i) * gap
    track.setFill()
    capsule(CGRect(x: left, y: y, width: barW, height: barH)).fill()
    item.0.setFill()
    capsule(CGRect(x: left, y: y, width: max(barH, barW * item.1), height: barH)).fill()
}

let cx: CGFloat = 778
let cy: CGFloat = size - 548
let r: CGFloat = 168
teal.setFill()
NSBezierPath(ovalIn: CGRect(x: cx - r, y: cy - r, width: r * 2, height: r * 2)).fill()
badgeFill.setFill()
let inner = r - 26
NSBezierPath(ovalIn: CGRect(x: cx - inner, y: cy - inner, width: inner * 2, height: inner * 2)).fill()

let check = NSBezierPath()
check.lineWidth = 46
check.lineCapStyle = .round
check.lineJoinStyle = .round
check.move(to: NSPoint(x: cx - 78, y: cy - 8))
check.line(to: NSPoint(x: cx - 18, y: cy - 68))
check.line(to: NSPoint(x: cx + 86, y: cy + 58))
teal.setStroke()
check.stroke()

image.unlockFocus()

try FileManager.default.createDirectory(at: dest.deletingLastPathComponent(), withIntermediateDirectories: true)
guard let tiff = image.tiffRepresentation,
      let rep = NSBitmapImageRep(data: tiff),
      let png = rep.representation(using: .png, properties: [:]) else {
    fputs("failed to encode AppIcon.png\n", stderr)
    exit(1)
}
try png.write(to: dest)
fputs("wrote \(dest.path)\n", stderr)
